### ISSUE TITLE
[Web] Uncaught ReferenceError: process is not defined with blank white screen

### ISSUE THREAD

**Issue by ramiel1999** [expo/expo#17547]
### Summary

Initializing fresh project with expo init, running expo start:web, and then changing App.tsx file to produce an error. For example: 
```
export default function App() {
  return (
    <View style={styles.container}>
      <Text>{hello.world()}</Text>
      <StatusBar style="auto" />
    </View>
  );
}
```
will show a blank white screen instead of an overlay of error from "react-error-overlay". As well as in console the following error:

```
Uncaught ReferenceError: process is not defined
    4043 blank line 4342 > injectedScript:2
    r blank line 4342 > injectedScript:2
    8048 blank line 4342 > injectedScript:2
    r blank line 4342 > injectedScript:2
    8641 blank line 4342 > injectedScript:2
    r blank line 4342 > injectedScript:2
    <anonymous> blank line 4342 > injectedScript:2
    <anonymous> blank line 4342 > injectedScript:2
    <anonymous> blank line 4342 > injectedScript:2
    onload index.js:1
```

### Managed or bare workflow? If you have `ios/` or `android/` directories in your project, the answer is bare!

managed

### What platform(s) does this occur on?

Web

### SDK Version (managed workflow only)

45

### Environment

```
  expo-env-info 1.0.3 environment info:
    System:
      OS: Windows 10 10.0.19044
    Binaries:
      Node: 16.14.0 - C:\Program Files\nodejs\node.EXE
      Yarn: 1.22.17 - ~\AppData\Roaming\npm\yarn.CMD
      npm: 8.3.1 - C:\Program Files\nodejs\npm.CMD
    IDEs:
      Android Studio: AI-212.5712.43.2112.8512546
    npmPackages:
      expo: ~45.0.0 => 45.0.4
      react: 17.0.2 => 17.0.2
      react-dom: 17.0.2 => 17.0.2
      react-native: 0.68.2 => 0.68.2
      react-native-web: 0.17.7 => 0.17.7
    Expo Workflow: managed
```

### Reproducible demo

# To Reproduce:

1. initialize new expo project: expo init project
2. in command line> expo start:web
3. change in app.tsx to create error. ex:` <Text>{hello.world()}<Text>` where hello is not defined
4. in browser, the screen should be blank with error in console about process is not defined

Github repo:
https://github.com/ramiel1999/expo-web-blank-screen

**Comment by Nebel69** [comment]
The error is as a result of react-error-overlay (which many people would never have heard of because it is a dependency of react-scripts). This package's dependencies were update to support webpack v5, which unfortunately is not compatible with react-scripts v4.

**Comment by ramiel1999** [comment]
I have seen some workarounds to resolve react-error-overlay's version, but shouldnt there be a definitive solution and not a workaround?

**Comment by tones31** [comment]
This happens to me on a very fresh install of expo following the expo steps
1. Install cli
2. Init a new project with expo-create-app or expo init

**Comment by fo-fo** [comment]
Why was this closed?

**Comment by deathemperor** [comment]
can confirm this is blocking my ejected expo app to run on web

**Comment by Jackman3005** [comment]
Can we get this opened back up or get an explanation as to why it should remain closed? We've been having this problem for quite some time and have struggled to even identify `react-error-overlay` as a part of the problem.

@itsramiel I'd like to hear about the workarounds that you've found to work.

**Comment by itsramiel** [comment]
@Jackman3005 unfortunately I was experimenting with rn web and did not really continue to work on it. Stuck on rn for ios/android for now.

Does it still not work with latest expo sdk though? Like a fresh new project with updated deps?

**Comment by itsramiel** [comment]
@Jackman3005  Yeah the error is still there, just checked. I really encourage you to open a new issue in my opinion

**Comment by Jackman3005** [comment]
I have spent some more time in this and finally got a fix for myself and more information on why it was happening.

### tl;dr
Force `react-error-overlay` to version `6.0.9` and it the process error will go away and the error screen will show again.

```json
"resolutions": {
  "react-error-overlay": "6.0.9"
}
```


### Why
Create React App owns the `react-error-overlay` and has recently updated all the `react-scripts` to support webpack 5. During this transition they pushed out an update to `react-error-overlay` to support webpack 5. Since it was only a patch version update most people have pulled the update into their code, but unfortunately it is not backwards compatible with Webpack 4!

[Expo support for webpack 5 is still in progress](https://github.com/expo/expo-cli/issues/2881)

[Issue with react-error-overlay and process not defined](https://github.com/facebook/create-react-app/issues/11773)

**Comment by itsramiel** [comment]
> I have spent some more time in this and finally got a fix for myself and more information on why it was happening.
> ### tl;dr
> 
> Force `react-error-overlay` to version `6.0.9` and the process not defined error will go away and the error screen will show again.
> 
> ```json
> "resolutions": {
>   "react-error-overlay": "6.0.9"
> }
> ```
> 
> ### Why
> 
> Create React App owns the `react-error-overlay` and has recently updated all the `react-scripts` to support webpack 5. During this transition they pushed out an update to `react-error-overlay` to support webpack 5. Since it was only a patch version update most people have pulled the update into their code, but unfortunately it is not backwards compatible with Webpack 4!
> 
> [Expo support for webpack 5 is still in progress](https://github.com/expo/expo-cli/issues/2881)
> 
> [Issue with react-error-overlay and process not defined](https://github.com/facebook/create-react-app/issues/11773)

if expo simply said this would have been better than silently closing the issue. Thanks @Jackman3005

**Comment by fo-fo** [comment]
Thanks @Jackman3005 for investigating it.

I noticed that in my case, quite annoyingly, `react-error-overlay` was actually being pulled from a globally installed `expo-cli` dependency. I hadn't noticed that it was using the globally installed version because I didn't expect `yarn expo` to work without having installed `expo-cli`. Long story short, I was able to fix it by adding `expo-cli` as a dev dependency and resolving `react-error-overlay` to version 6.0.9.

I was planning to use `patch-package` to patch the `process` access in `react-error-overlay` to something like `window.process || {}` to circumvent the error, but since they distribute minified (single-line) source in the npm package that would produce a 745 KiB patch. So I guess I'm not doing that.

**Comment by orta** [comment]
Confirming that Expo 46 still gets this behavior out of the box also, solved via "resolutions" above

From Expo's perspective this is a tricky one as its a transitive dependency. Perhaps this should be (systemically) solved by defining `process` when in dev mode in the [default webpack config](https://github.com/expo/expo-cli/blob/main/packages/webpack-config/src/webpack.config.ts)?

**Comment by PerryRylance** [comment]
Why is this closed? Seeing this on a fresh Expo build when trying to launch for web...

**Comment by hugobarragon** [comment]
Reopen, this still happens on SDK46

**Comment by stevus** [comment]
Needs to be opened back up - I'm going to create an issue of my own since I'm actively blocked by this.

Hey @Jackman3005, how do I implement this? Not sure what I'm supposed to here - 

> I have spent some more time in this and finally got a fix for myself and more information on why it was happening.
> 
> ### tl;dr
> Force `react-error-overlay` to version `6.0.9` and the process not defined error will go away and the error screen will show again.
> 
> ```json
> "resolutions": {
>   "react-error-overlay": "6.0.9"
> }
> ```
> 
> ### Why
> Create React App owns the `react-error-overlay` and has recently updated all the `react-scripts` to support webpack 5. During this transition they pushed out an update to `react-error-overlay` to support webpack 5. Since it was only a patch version update most people have pulled the update into their code, but unfortunately it is not backwards compatible with Webpack 4!
> 
> [Expo support for webpack 5 is still in progress](https://github.com/expo/expo-cli/issues/2881)
> 
> [Issue with react-error-overlay and process not defined](https://github.com/facebook/create-react-app/issues/11773)

**Comment by stevus** [comment]
OK I've added `"react-error-overlay": "6.09"` to package.json as a devDependency as outlined [here](https://github.com/facebook/create-react-app/issues/11773#issuecomment-995933869).

However, I'm still blocked on application dev:

![image](https://user-images.githubusercontent.com/1261603/208888674-9acc5a62-dc2f-4e45-80ff-5a14fd75fe46.png)

**Comment by Jackman3005** [comment]
Hey @stevus sorry for the delay in getting back to you. Is it possible you are using NPM as your package manager? I updated my [original answer](https://github.com/expo/expo/issues/17547#issuecomment-1235148185) to better explain and include a link for how to do it with NPM.

You can tell if what you're doing works by checking your `yarn.lock`(if using `yarn`) or `package-lock.json` (if using NPM) and seeing exactly which version of `react-error-overlay` is being used. It needs to be less than or equal to`6.0.9` as the problematic code was introduced in version `6.0.10`.

Sometimes your bundler will need to be restarted if its not picking up the changed package and restarting on its own.

**Comment by bradjones1** [comment]
On SDK 47 here, I was experiencing a situation where I had an iframe covering the app (over my ErrorBoundary fallback component) which was blocking interaction with the error boundary (e.g., to reset it.) I tried initially setting the `resolutions` key but had to explicitly specify `6.0.9` as a dev dependency. As noted above, it must be pinned to this version, not the latest (which at this time is `6.0.11`.

I am still refining my JS skill so hot reloading, webpack and the like are still a bit of black magic to me... so I'm just happy to have it working. The linked issue on CRA, etc. seem to be full of unhappy customers so I'm not sure what's going on more broadly. But in any case, this is what worked for me. Previously the version of `react-error-boundary` was 5.x, as a result of Expo's dependencies. So I imagine there's room for improvement on Expo's side to require a known-good version in the meantime while an upstream fix gets hashed out?

**Comment by rodolfobarretoweb** [comment]
I'm having the same problem using expo 48 and unfortunately nothing that everyone indicated to fix the issue works.

**Comment by howardwu1** [comment]
I'm curious if we are forced to resolve it via npm rather than yarn now since the expo-cli is deprecated now and uses npx expo instead. I'll try to do it for npm and report back

**Comment by finnmerlett** [comment]
I'll just leave my issue and resolution in case anyone finds this with the same problem as me. The "process is not defined" error was being caused by `babel-plugin-inline-dotenv` substituting out a `process.env.ENV_VAR` to be `process && [...]`. Removing this babel plugin fixed the issue, I realised I didn't even need it as I had a webpack DefinePlugin doing .env replacement already, and in a less fragile way.

**Comment by finnmerlett** [comment]
Oh and if anyone needs the `EXPO_PUBLIC_<...>` env variables added in to the process.env substitution carried out by the config from `createExpoWebpackConfigAsync` (since it doesn't appear to happen automatically at this point with Expo 49), here's how to patch them into the webpack.config.js:

```javascript
const createExpoWebpackConfigAsync = require("@expo/webpack-config");
const { DefinePlugin } = require("webpack");
const get = require("lodash/get");

// Expo CLI will await this method so you can optionally return a promise.
module.exports = async function configureWebpack(
  /** @type {import("@expo/webpack-config/webpack/types").InputEnvironment} */ env,
  /** @type {import("@expo/webpack-config/webpack/types").Arguments} */ argv,
) {
  // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
  const config = await createExpoWebpackConfigAsync(env, argv);

  // Add in missing process.env.EXPO_PUBLIC_<...> variables to the DefinePlugin
  const expoPublicEnvVars = Object.fromEntries(
    Object.entries(process.env)
      .filter(([key]) => key.startsWith("EXPO_PUBLIC_"))
      .map(([key, value]) => [key, JSON.stringify(value)]),
  );
  const configDefinePlugin = config.plugins?.[4];
  if (!(configDefinePlugin instanceof DefinePlugin)) {
    throw new Error("unexpected config plugins array order");
  }
  const configDefinePluginEnv = /** @type {{}} */ (
    get(configDefinePlugin.definitions, ["process.env"])
  );
  Object.assign(configDefinePluginEnv, expoPublicEnvVars);

  // Finally return the new config for the CLI to use.
  return config;
};

```

You can swap out the lodash.get for something simpler and less type-safe if you aren't trying to play nice with a ts type-checker. Also you may need to drop a breakpoint in to check which array item the existing DefinePlugin is occupying. For me it was index 4, it may be different for you.
